superset/connectors/sqla/models.py

```diff
@@ -175,10 +175,11 @@ def dttm_sql_literal(self, dttm):
         if self.database_expression:
             return self.database_expression.format(dttm.strftime('%Y-%m-%d %H:%M:%S'))
         elif tf:
+            seconds_since_epoch = int(dttm.timestamp())
```

Comment:
Are we sure that `int(dttm.timestamp()) == (dttm - datetime(1970, 1, 1)).total_seconds()` ? Absolutely and always? Timezones included?

Reply:
On the subject of timezones, there is a bit of a fundamental issue here: The timestamp column can be in one TZ, while superset-webserver itself is running in another TZ. And then the UI also allows entering timestamp filters, which could be in another TZ. 

My thought was that perhaps we should allow the user to specify what the TZ of the timestamp column is, particularly if the timestamp is in a numeric format like epoch_s. And then honor that. 

Currently, doing "dttm - datetime(1970, 1, 1)" is problematic because dttm could be in TZ like UTC, while datetime(1970, 1, 1) is in the superset-webserver's TZ (which could be PST). So this calculation of UTC - PST isn't well defined.

OTOH, dttm.timestamp() returns the seconds since epoch relative to the dttm's own TZ (which if none is UTC, else whatever the dttm was created with).

Both approaches have their shortcomings, and I think the only proper fix is to have superset become more TZ aware.

Reply:
I agree that we need a revamp of timezone handling in Superset, though we shouldn't break how current charts display time multiple times on the way there.

I know when using datasources with timestamp, that stuff is fairly clear/straightforward, meaning there's really just the time of the event and the time of the user's browser to take into consideration.  For other things like a truncated `ds` or for example our "baby birth names" dataset, where the data points are on the first day of the year, we often don't want any form of conversion. Like I don't what the chart to adjust for timezones at all or people in different timezones will see something like `1/1/x 9AM PST`.

There are thousands of charts in the wild that have adapted (perhaps using time expressions, or the datasource time lag feature, or even altered pipelines and source data) to whatever the current behavior is now in Superset and we should not make changes that will alter people's existing charts. This requires some thinking and a "Superset Improvement Proposal", and change management. Probably a per-chart feature flag around using the "new and improved time handling" that would be False for existing charts and True for new ones.

Reply:
I think this method does not affect anything about how the timestamp values are rendered in the viz. It only controls how the filter is rendered. Ie how do we want to convert the "since"/"until" values in the form to the SQL timestamp comparison filter. 

The current behavior is as follows:
1. If the user gives the start/end with timestamp: (a) If the DB column is epoch_s, then this will throw an error since we are doing aware-dttm minus naive-dttm, which is an error. (b) If the DB column is not epoch, then the user specified time will be converted into the timestamp string as given by the column's timestring format. 
2. If the user entered start/end is without timestamp, as in the majority of cases: (a) If the DB columns is epoch_s, then we will return the number of seconds since 1/1/1970 (local time). (b) Otherwise, the timestamp would be converted using python's strftime.

epoch_s/epoch_ms lack information about what timezone is the "epoch" calculation done using. Technically it should be UTC, but users could be subtracting it from 1/1/1970 in their local time zone (and using the time-shift feature instead).

Incorporating timezone in the database expression is easy: The user can provide an '%z' if the underlying DB supports timezones in literals.

I agree with you on the need to preserve absolute backward compatibility and as such I will redo my implementation:

- I will introduce a epoch_ms_tz/epoch_s_tz: Which means that the DB column is in number of milliseconds/seconds since 1/1/1970 (given TZ). If tz is not given, then the default is to make it behave like today: time delta since 1/1/1970 (local TZ).

[ My current implementation was changing the cases 1.a and 2.a above ]

Reply:
Actually, I think dttm.timestamp() is the right thing to do here: 

The database functions epoch_to_dttm already convert the expression to the unix timestamp. So the assumption is already that epoch_s means UTC-epoch_s. So this function dttm_sql_literal converts the control input start/end timestamps to epoch-time. And I would argue that it too should behave the same way. 

I can feature flag both these changes since they are "breaking" but I think the right thing to do. 

I decided against introducing a new kind of epoch_s_tz because the db_engine_specs.py file already assumes that epoch_s is in UTC epoch (like it should).